picotest_helpers/src/migration.rs

```diff
@@ -0,0 +1,473 @@
+where
+    P: AsRef<Path>,
+{
+    let Some(file_name) = file_name.as_ref().file_name() else {
+        bail!("migration file does not have file name")
+    };
+    let Some(file_name) = file_name.to_str() else {
+        bail!("migration file have non-utf8 name")
+    };
+    let Some((name, ext)) = file_name.rsplit_once('.') else {
+        bail!("migration file does not have an extension")
+    };
+    if ext.to_lowercase() != "sql" {
+        bail!("migration file does not have sql extension")
+    }
+    let Some((version, migration_name)) = name.split_once('_') else {
+        bail!("migration file has invalid name")
+    };
+    let Ok(version) = MigrationVersion::from_str(version) else {
+        bail!("failed to parse migration version: {version}")
+    };
+    Ok((version, migration_name.to_string()))
+}
+
+pub fn parse_migration_text<S>(sql_text: S) -> Result<Vec<MigrationStatement>, Error>
+where
+    S: AsRef<str>,
+{
+    let sql_text = sql_text.as_ref();
+    let mut output = Vec::with_capacity(sql_text.matches('\n').count());
+    let mut acc = Option::<String>::None;
+
+    for line in sql_text.lines() {
+        // skip empty lines
+        let prefix_line = line.trim_end();
+        if prefix_line.is_empty() {
+            continue;
+        }
+        let line = prefix_line.trim_start();
+        // single line comment
+        if line.starts_with("--") {
+            // ignore if currently building a statement
+            if acc.is_none() {
+                output.push(MigrationStatement::new(line));
+                continue;
+            }
+        }
+        // append and insert statement text
+        if let Some(acc_string) = acc.take() {
+            acc = Some(acc_string + line)
+        } else {
+            acc = Some(String::from(line));
+        }
+        // statement was not finished, continue building
+        if !line.ends_with(';') {
+            continue;
+        }
+        let acc_string = acc.take().unwrap();
+        output.push(MigrationStatement::new(acc_string));
+    }
+    Ok(output)
+}
+
+/// Extract indexes [a,b), where starts and ends migrations by type.
+#[allow(clippy::type_complexity)]
+fn extract_up_down_ranges(
+    statements: &[MigrationStatement],
+) -> Result<((usize, usize), (usize, usize)), Error> {
+    let mut up_range_start = 0;
+    let mut down_range_start = 0;
+    let end_range = statements.len();
+    for (idx, statement) in statements.iter().enumerate() {
+        if statement.is_line_comment() {
+            if statement.text().starts_with("-- pico.UP") {
+                up_range_start = idx
+            }
+            if statement.text().starts_with("-- pico.DOWN") {
+                down_range_start = idx
+            }
+        }
+    }
+    Ok((
+        (up_range_start, down_range_start),
+        (down_range_start, end_range),
+    ))
+}
+
+pub fn parse_migration_file<P>(path: P) -> Result<Migration, Error>
```

Comment:
Давай его тоже спрячем, пусть только parse_migrations торчит наружу.

```suggestion
fn parse_migration_file<P>(path: P) -> Result<Migration, Error>
```